src/clj/puppetlabs/services/jruby_pool_manager/impl/instance_pool.clj

```diff
@@ -0,0 +1,65 @@
+(ns puppetlabs.services.jruby-pool-manager.impl.instance-pool
+  (:require [puppetlabs.services.jruby-pool-manager.impl.jruby-agents :as jruby-agents]
+            [puppetlabs.services.protocols.jruby-pool :as pool-protocol]
+            [puppetlabs.services.jruby-pool-manager.impl.jruby-internal :as jruby-internal]
+            [puppetlabs.services.jruby-pool-manager.jruby-schemas :as jruby-schemas]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n])
+  (:import (puppetlabs.services.jruby_pool_manager.jruby_schemas InstancePool)))
+
+(extend-type InstancePool
+  pool-protocol/JRubyPool
+  (fill
+    [this]
+    (let [modify-instance-agent (jruby-agents/get-modify-instance-agent this)]
+      (jruby-agents/send-agent modify-instance-agent
+                               #(jruby-agents/prime-pool! this))))
+
+  (shutdown
+    [this]
+    (jruby-agents/flush-pool-for-shutdown! this))
+
+  (lock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lock pool)))
+
+  (lock-with-timeout
+    [this timeout time-unit]
+    (let [pool (jruby-internal/get-pool this)]
+      (.lockWithTimeout pool timeout time-unit)))
+
+  (unlock
+    [this]
+    (let [pool (jruby-internal/get-pool this)]
+      (.unlock pool)))
+
+  (borrow
+    [this]
+    (jruby-internal/borrow-from-pool this))
+
+  (borrow-with-timeout
+    [this timeout]
+    (jruby-internal/borrow-from-pool-with-timeout this timeout))
+
+  (return
+    [this instance]
+    (when (jruby-schemas/jruby-instance? instance)
+      (let [new-state (swap! (jruby-internal/get-instance-state-container instance)
+                             update-in [:borrow-count] inc)
```

Comment:
Should `update-in [:borrow-count] inc` be in its own set of parentheses? Maybe it doesn't need to be, but I think that would help me read this method? Unless I'm misunderstanding what `swap!` does..

Reply:
The lack of parens is [required](https://clojuredocs.org/clojure.core/swap!). `swap!` takes several args, first the atom being updated, then a function, then each of the args to that function, as many as there are. In this case, the function is `update-in`, which itself takes the map (passed in by `swap!`), the list of keys pointing to the value to be udpated, then the function to apply to the value of that key.

It's honestly pretty confusing. But in general, no parens are ever optional in Clojure. If they're there, they tend to be there for a reason, same if they're missing. There's very little syntax leeway in the language.

Reply:
ahh good to know, thanks!

Reply:
I think this is idiomatic (and the way it was originally written [here](https://github.com/puppetlabs/jruby-utils/pull/78/files#diff-63fb42862df919c7a769bbd0da1c3ee3L349)), though could also be:
```clj
(swap! ... #(update-in % [:borrow-count] inc))
```

Reply:
Oh huh. I do like that better, I'll update.